server/svix-server/src/core/security.rs

```diff
@@ -77,11 +128,89 @@ where
     }
 }
 
+pub struct AuthenticatedOrganization {
+    pub permissions: Permissions,
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for AuthenticatedOrganization
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self> {
+        let permissions = Permissions::from_request(req).await?;
+        match permissions.type_ {
+            KeyType::Organization => {}
+            KeyType::Application => {
+                return Err(HttpError::unauthorized(
+                    None,
+                    Some(
+                        "You are only permitted to perform operations under the Application type"
+                            .to_owned(),
+                    ),
+                )
+                .into());
+            }
+        }
+
+        Ok(AuthenticatedOrganization { permissions })
+    }
+}
+
 #[derive(Deserialize)]
 struct ApplicationPathParams {
     app_id: ApplicationIdOrUid,
 }
 
+pub struct AuthenticatedOrganizationWithApplication {
+    pub permissions: Permissions,
+    pub app: application::Model,
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for AuthenticatedOrganizationWithApplication
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self> {
+        let permissions = Permissions::from_request(req).await?;
+
+        match permissions.type_ {
+            KeyType::Organization => {}
+            KeyType::Application => {
+                return Err(HttpError::unauthorized(
+                    None,
+                    Some(
+                        "You are only permitted to perform operations under the Application type"
+                            .to_owned(),
+                    ),
+                )
```

Comment:
It should be a `permission_denied` (so 403) not a 401.
And just say "insufficient access".